### PR TITLE
fix: align JSON types with official ETSI TS 119 602-1 schema

### LINKED ISSUE #28: JSON schema non-conformance: types must match ETSI TS 119 602-1 official JSON schema
## Problem

The current JSON serialization format in `pkg/etsi119602` does not conform to the official ETSI TS 119 602-1 JSON schema (`1960201_json_schema.json` from the ETSI forge).

### Structural
- **Missing root wrapper**: Official schema requires `{"LoTE": {...}}` root object
- **Wrong top-level key**: Official uses `ListAndSchemeInformation`, we use `schemeInformation`
- **Pointers location**: Official has `PointersToOtherLoTE` *inside* `ListAndSchemeInformation`, we have `pointersToOtherLoTEs` at top level
- **Entity structure**: Official nests into `TrustedEntityInformation` + `TrustedEntityServices`, we use a flat structure

### Field naming (official PascalCase vs current camelCase/custom)
| Official | Current |
|---|---|
| `LoTEVersionIdentifier` (int) | `version` (string) |
| `LoTESequenceNumber` | `sequenceNumber` |
| `LoTEType` | `schemeType` |
| `SchemeOperatorName` | `schemeOperator` |
| `SchemeTerritory` | `territory` |
| `ListIssueDateTime` | `issueDate` |
| `NextUpdate` | `nextUpdate` |
| `PointersToOtherLoTE` | `pointersToOtherLoTEs` |
| `LoTELocation` | `location` |
| `TrustedEntitiesList` | `trustedEntities` |
| `TEName` | `entityName` |

### Multi-language types
- Official: `{"lang": "en", "value": "..."}` and `{"lang": "en", "uriValue": "..."}`
- Current: `{"language": "en", "value": "..."}` and `{"language": "en", "uri": "..."}`

### Pointer qualifiers
- Official: Each `OtherLoTEPointer` has `LoTEQualifiers` array containing `{LoTEType, SchemeOperatorName, SchemeTerritory, MimeType}`
- Current: Flat `schemeType` and `schemeTerritory` on the pointer directly

### Digital identities
- Official: `ServiceDigitalIdentity` with `X509Certificates`, `PublicKeyValues` (JWK per RFC 7517), `X509SubjectNames`, `X509SKIs`, `OtherIds`
- Current: Flat `DigitalIdentity` with `type` discriminator and `jwk`/`x509Certificate`/`did` fields

## Impact

- Cannot consume JSON from other ETSI TS 119 602-1 implementations (e.g. publish-trusted-list tooling)
- Cannot produce JSON that validates against the official schema
- go-trust LoTE registry will break when fed official-format JSON

## Plan

Rewrite the types in `pkg/etsi119602` to match the official schema exactly, keeping XML support working via separate XML struct tags. Update all consumers (go-trust, trust-lists).

### PR DESCRIPTION
**Pull request by leifj**
## Summary

Rewrites the JSON serialization types in `pkg/etsi119602` to conform to the official ETSI TS 119 602-1 JSON schema.

### Changes

- **lote.go**: Restructured all types to match the official schema — `LoTEDocument`, `ListOfTrustedEntities`, `ListAndSchemeInformation`, `TrustedEntity`, `ServiceInformation`, `ServiceDigitalIdentity`, etc.
- **lotl.go**: Updated `ListOfTrustedLists` type alias and LoTL-specific constants/functions to use the new types.
- **fetch.go**: Aligned fetch functions with new type signatures.
- **Tests**: Added tests documenting IDUnion XML non-conformances with the TS 119 602-1 XSD.

### Breaking Changes

This is a breaking change to the Go types. Consumers (e.g. `go-trust`) must update to use the new field names and type structure.

Closes #28

### CODE REVIEW

File: pkg/etsi119602/xml.go [hunk 1]
```diff
@@ -471,33 +441,36 @@ func xmlToTrustedEntity(te *xmltypes.TEType) TrustedEntity {
 
 // --- Service conversion ---
 
-func entityServiceToXML(s EntityService) *xmltypes.TrustedEntityServiceType {
+// entityServiceToXMLTE is unused placeholder to avoid compile error from earlier code
+func entityServiceToXMLTE(_ TrustedEntityService) *xmltypes.TEType {
+	return nil
+}
+
```
[the remark this answers is not in the document]
Comment (leifj) [reply pkg/etsi119602/xml.go]: Fixed in 8be6f07 — removed the `entityServiceToXMLTE` stub and the unused `svcs` variable. The `trustedEntityToXML` function now only builds services via `entityServiceToXML`.

File: pkg/etsi119602/lote.go [hunk 2]
```diff
@@ -204,20 +218,17 @@ const (
...
 func ParseLoTE(data []byte) (*ListOfTrustedEntities, error) {
-	var lote ListOfTrustedEntities
-	if err := json.Unmarshal(data, &lote); err != nil {
+	var doc LoTEDocument
+	if err := json.Unmarshal(data, &doc); err != nil {
 		return nil, fmt.Errorf("failed to parse LoTE: %w", err)
 	}
-	return &lote, nil
+	return &doc.LoTE, nil
 }
```
[the remark this answers is not in the document]
Comment (leifj) [reply pkg/etsi119602/lote.go:238]: Fixed in 8be6f07 — ParseLoTE now validates that the unmarshalled document is not a zero-value (checks SchemeOperatorName, LoTEVersionIdentifier, and TrustedEntitiesList) and returns an error if the LoTE root key is missing.

File: pkg/etsi119602/fetch_test.go [hunk 3]
```diff
@@ -12,54 +12,37 @@ import (
...
+	data, err := lote.MarshalIndent()
+	require.NoError(t, err)
 	require.NoError(t, os.WriteFile(path, data, 0644))
 
-	result, err := FetchLoTE(path, nil)
+	fetched, err := FetchLoTE(path, nil)
 	require.NoError(t, err)
-	assert.Equal(t, "SE", result.SchemeInformation.Territory)
+	assert.Equal(t, "SE", fetched.ListAndSchemeInformation.SchemeTerritory)
 }
```
[the remark this answers is not in the document]
Comment (leifj) [reply pkg/etsi119602/fetch_test.go:27]: Acknowledged — additional error-path tests for FetchLoTE/FetchRaw (HTTP status codes, invalid JSON/XML, unreachable host, Content-Type handling) are tracked for a follow-up. The current tests cover the core happy path and basic validation.

File: pkg/etsi119602/xml_test.go [hunk 4]
```diff
@@ -2,442 +2,157 @@ package etsi119602
...
-	assert.Contains(t, xmlStr, "TrustedEntityInformation")
-	assert.Contains(t, xmlStr, "TrustedEntityServices")
-	assert.Contains(t, xmlStr, "Swedish Authority")
-	assert.Contains(t, xmlStr, "credential-issuer")
-	assert.Contains(t, xmlStr, "LOTETag")
+	assert.Contains(t, string(xmlData), "ListOfTrustedEntities")
 
-	// Parse back
 	parsed, err := ParseLoTEXML(xmlData)
 	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (leifj) [reply pkg/etsi119602/xml_test.go:21]: Acknowledged — XML error-path tests (malformed XML, empty/truncated input, edge cases around IDUnion/lenient parsing) are tracked for a follow-up.

File: pkg/etsi119602/lote_test.go [hunk 5]
```diff
@@ -2,146 +2,235 @@ package etsi119602
...
 
-	// Parse back
+	// Verify root wrapper
+	var raw map[string]json.RawMessage
+	require.NoError(t, json.Unmarshal(data, &raw))
+	_, hasLoTE := raw["LoTE"]
+	assert.True(t, hasLoTE, "JSON root must have LoTE key")
+
 	parsed, err := ParseLoTE(data)
 	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (leifj) [reply pkg/etsi119602/lote_test.go:66]: Fixed in 8be6f07 — added TestParseLoTE_MissingRootKey which tests unwrapped JSON, empty objects, and invalid JSON.

File: pkg/etsi119602/xml.go [hunk 6]
```diff
@@ -506,187 +479,114 @@ func entityServiceToXML(s EntityService) *xmltypes.TrustedEntityServiceType {
...
-	if xid.X509SubjectName != nil {
-		return DigitalIdentity{
-			Type:            "x509_subject_name",
-			X509SubjectName: *xid.X509SubjectName,
-		}, true
+	for _, name := range sdi.X509SubjectNames {
+		x.DigitalId = append(x.DigitalId, &xmltypes.DigitalIdentityType{
+			X509SubjectName: &name,
+		})
 	}
```
[the remark this answers is not in the document]
Comment (leifj) [reply pkg/etsi119602/xml.go:532]: Fixed in 8be6f07 — the range loop now copies cert.Val and name into local variables before taking their address.

File: pkg/etsi119602/xml.go [hunk 7]
```diff
@@ -271,198 +229,210 @@ func xmlNamesToNameSet(x *xmltypes.InternationalNamesType) NameSet {
...
+			if q.LoTEType != "" {
+				lang := xmltypes.Lang("en")
+				val := xmltypes.NonEmptyString(q.LoTEType)
+				xq.TextualInformation = append(xq.TextualInformation, &xmltypes.MultiLangStringType{
+					XmlLangAttr:    &lang,
+					NonEmptyString: &val,
+				})
+			}
+			xp.LoteAdditionalInformation = xq
+		}
```
[the remark this answers is not in the document]
Comment (leifj) [reply pkg/etsi119602/xml.go:349]: Fixed in 8be6f07 — refactored to build a single AdditionalInformationType and append all qualifier TextualInformation entries into it, rather than overwriting on each iteration.

File: pkg/etsi119602/xml.go [hunk 8]
```diff
@@ -271,198 +229,210 @@ func xmlNamesToNameSet(x *xmltypes.InternationalNamesType) NameSet {
...
+		svcs := &xmltypes.TrustedEntitiesListType{}
+		for _, s := range e.TrustedEntityServices {
+			svcs.LoteTrustedEntity = append(svcs.LoteTrustedEntity, entityServiceToXMLTE(s))
+		}
+		// Use the proper field for services
+		svcList := &xmltypes.TrustedEntityServicesListType{}
+		for _, s := range e.TrustedEntityServices {
+			svcList.LoteTrustedEntityService = append(svcList.LoteTrustedEntityService, entityServiceToXML(s))
+		}
+		te.LoteTrustedEntityServices = svcList
```
[the remark this answers is not in the document]
Comment (leifj) [reply pkg/etsi119602/xml.go:413]: Fixed in 8be6f07 — removed the intermediate svcs variable and the entityServiceToXMLTE stub. Services are now built directly via entityServiceToXML into TrustedEntityServicesListType.

**Comment by AniaAlex** [comment]
## Question about schema compliance

```go
type TrustedEntityInformation struct {
    TEName                  NameSet                `json:"TEName"`
    TETradeName             NameSet                `json:"TETradeName,omitempty"`
    TEAddress               *TEAddress             `json:"TEAddress,omitempty"`      // required per schema?
    TEInformationURI        []NonEmptyMultiLangURI `json:"TEInformationURI,omitempty"` // required per schema?
    TEInformationExtensions []json.RawMessage      `json:"TEInformationExtensions,omitempty"`
}
```

Per the official JSON schema, `TEAddress` and `TEInformationURI` are **required** in `TrustedEntityInformation`:
```json
"required": ["TEName", "TEAddress", "TEInformationURI"]
```

Should remove `omitempty` from these fields?   just interesting that: ```Country```in json and ```CountryName``` in xml , but in schema it seems to be like this

**Comment by leifj** [comment]
Good catch! Fixed in d15b45b:

- Removed `omitempty` from `TEAddress` and `TEInformationURI` in the struct tags
- Added validation checks for both required fields in `validate.go`
- Updated all test fixtures to include the required address and URI data

Re `Country` vs `CountryName`: correct, the JSON schema uses `Country` (and our JSON struct does too), while the XML schema uses `CountryName` (and our XML types use that). Both are correct for their respective formats.

**Review by AniaAlex — APPROVED** [review]
